ksonnet.beta.3/k8s.libsonnet

```diff
@@ -319,6 +319,7 @@
       deployment:: {
         local kind = {kind: "Deployment"},
         new(name, replicas, containers, podLabels={app: name}):: apiVersion + kind + self.mixin.metadata.withName(name) + self.mixin.spec.withReplicas(replicas) + self.mixin.spec.template.spec.withContainers(containers) + self.mixin.spec.template.metadata.withLabels(podLabels),
+        newWithoutReplicas(name, containers, podLabels={app: name}):: apiVersion + kind + self.mixin.metadata.withName(name) + self.mixin.spec.template.spec.withContainers(containers) + self.mixin.spec.template.metadata.withLabels(podLabels),
```

Comment:
@glindste Hi, good catch! I also believe this should be fixed somehow.

However, assuming `replicas` isn't actually a required parameter for a deployment, shouldn't we just remove `replicas` from the original constructor instead?

Then, you can still explicitly set `replicas` like:

```
deployment.new(
  params.name,
  params.replicas,
  container
    .new(params.name, params.image)
    .withPorts(port.new(params.port)),
  {app: params.name}
).withReplicas(2)
```

Reply:
Yeah good point. I wouldn't mind removing replicas from the default constructor since it's not a required parameter. If the maintainers agree I'll change the PR to modify the default constructor instead.

Reply:
I've been thinking about how to handle constructors. We can proceed in a couple of ways. 

1. Only include required items in constructors.
1. Employ a builder pattern for everything.

Reply:
Also, the `withoutX` may be too difficult to conceptualize.

Reply:
I would argue for option 1. Constructors are convenient in that they let the user know what are the minimum requirements for a valid resource definition, and then they can just add on everything else with builders. And if anyone wants to do pure builder pattern you can just start off with `{kind: "Deployment"}` or equivalent and add onto that. But I think it's valuable that constructors only have the required parameters, because consistency is important for the usability of the library. If there's a mix of constructors with only the required params and some with added ones that are optional in reality it can become confusing, especially when cross-referencing with the kubernetes api spec.